temporal-sdk/src/main/java/io/temporal/worker/WorkflowTaskDispatchHandle.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker;
+
+import io.temporal.api.workflowservice.v1.PollWorkflowTaskQueueResponse;
+import io.temporal.internal.worker.WorkflowTask;
+import java.io.Closeable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Semaphore;
+import java.util.function.Consumer;
+
+public class WorkflowTaskDispatchHandle implements Closeable {
+  private final CompletableFuture<Void> completed = new CompletableFuture<>();
+  private final Consumer<WorkflowTask> dispatchCallback;
+  private final Semaphore executorSlotsSemaphore;
+
+  public WorkflowTaskDispatchHandle(
+      Consumer<WorkflowTask> dispatchCallback, Semaphore executorSlotsSemaphore) {
+    this.dispatchCallback = dispatchCallback;
+    this.executorSlotsSemaphore = executorSlotsSemaphore;
+  }
+
+  /**
+   * @param workflowTask to be fed directly into the workflow worker
+   * @throws IllegalArgumentException if the workflow task doesn't belong to this worker
+   */
+  public void dispatch(PollWorkflowTaskQueueResponse workflowTask) {
```

Comment:
Would be curious to know what situations dispatch is not called. Arguably (and maybe this is in favor of the no-interceptor argument), the building of this handle and the invocation of dispatch are so close together that every handle is expected to call dispatch and this doesn't have to be closeable.

Reply:
It's a normal scenario where it's not called:
1. We obtain the handler
2. Server doesn't make an eager dispatch
3. We call `close` and don't call `dispatch`.

Reply:
Ah, duh, ok. I wonder if this is clearer:

```java
public void resolve(@Nullable PollWorkflowTaskQueueResponse workflowTask) {
```

Just a bit concerned that there isn't a single clear call to this unfortunately-user-facing class. But this does work as is today.

Reply:
I can move it into an internal package to show that users typically shouldn't be working with it.

> Just a bit concerned that there isn't a single clear call to this unfortunately-user-facing class. 
I'm not sure. I like the current structure. We call dispatch only if there is something to dispatch. We call close always. It makes the class safe to use and the eager path is distinct, which is good for readability.

When Local Completion is implemented and this method will start to return a LocalCompleteion instance, it will become even more clear that it's good to have a separate eager-path.